Added configuration to allow CVT to oversample from larger tiles

beaudet opened the pull request:
Added OVERSAMPLING_FACTOR as a configurable option (defaults to 1) with permitted values 1,2,3,4.   When > 1, CVT will attempt to use a tile size that is N times as large as the requested output. This generally leads to higher image quality in the output tile regardless of the resampling algorithm used although depending on the resampling algorithm, this might create more aliasing / sharpness than is desirable.  The sweet spot for the bilinear filter seems to be around 2.  The end user should experiment with combinations of resampling filters and the oversampling factor to achieve the desired quality.

beaudet commented:
standard (default) factor = 1 and bilinear (default) resampling algorithm
![image](https://cloud.githubusercontent.com/assets/77699/25859385/61317ed4-34ac-11e7-8656-e984d6d45cbf.png)


oversampled with factor = 2 and bilinear (default) resampling algorithm
![image](https://cloud.githubusercontent.com/assets/77699/25859398/6c4217c0-34ac-11e7-9bbf-9ce7591535cf.png)

oversampled with factor = 4 and bilinear (default) resampling algorithm
![image](https://cloud.githubusercontent.com/assets/77699/25859411/7257376c-34ac-11e7-8827-3059236a51dd.png)

ahankinson commented:
Is there a noticeable performance impact on higher values?

ruven commented:
In image processing theory, there shouldn't be any improvement beyond factor=2 (assuming the pyramid has been generated correctly). @beaudet are you seeing something something noticeable as you increase this factor? Don't forget also that each extra factor multiplies by 4 the raw time and memory required to do this processing.

beaudet commented:
Not surprisingly there is performance impact.  In a simulation involving "typical" source files and tile sizes, I observe the following differences - these samples are mostly grayscale sources from LOC. 

OVERSAMPLING FACTOR                              OUTPUT JPG TILES/SEC
1                                                                     122
2                                                                     84
3                                                                     70
4                                                                     70

I suspect the last result is due to most of the 4x the requested tile sizes not existing in the source images, so I've repeated with much larger source images and requests for consistent 400 pixel bounding box output tiles.  Most of those sources have images > 1600 pixels so this should give a better idea of the impact on performance.  

OVERSAMPLING FACTOR                              OUTPUT JPG TILES/SEC
1                                                                     75
2                                                                     36
3                                                                     23
4                                                                     11

And for requests for 200 pixel bounding box,:
1                                                                     133
2                                                                     84
3                                                                     63
4                                                                     39

So, the rate of performance impact also has to due with the size of the tile requested, but I'd say in general that it seems to scale somewhat linearly with the oversampling factor N as opposed to say, with the square of the scaling factor which logically I would expect to find as N gets really large and assuming very large sample images.

beaudet commented:
There does seem to be more sharpness at levels 3 and 4 than at level 2 although some of that is undesirable in my opinion.  After I merge the Lanczos filter in my local branch, I'll reassess with that filter.  If there's no desirable improvement there, it would probably make sense to allow a double N to be defined where 1 <= N <= 2 if the theory is consistent with the implementation of the filters..

ruven commented:
Thanks

beaudet commented:
I've merged all my enhancements and added ICC support to PNG as well which I'll have to submit a PR for once Ruven's develop branch progresses further.   I realized the way I implemented the oversampling is insufficient and it wasn't producing the intended resampling.  I re-implemented and now I see what you're saying.  The differences between using a sample 2x as large and 4x as large are pretty much non-existent.  However, there IS is big difference (quite noticeable) from 1x and 2x.   I will change the feature to use a float that must be between 1.0 and 2.0.  That will allow the calculated tile size to be used in cases where the resolution provides tiles that are already f * w greater than the requested size but will force a larger tile to be used when insufficient.  I'm interested in seeing the difference between 1.5 and 2.0.

I also noted that with JPEG format, the colors get washed out in cases where not many pixels are present.  This difference can seem quite noticeable when viewing PNG vs. JPEG images in the browser and zooming in at 500%.  However, at 100% these differences become almost negligible.  I suppose that's the compromise inherent with JPEG over a lossless format.

beaudet commented:
Seems like 1.5 is a reasonable value.  The problematic cases seem to be where the requested size is nearly as large as the tile size used for the sampling.  1.5 ensures that the tile being used is at least 50% larger.  1.25 is almost as good actually, but I think the quality differences start to become noticeable around that threshold.  I'll probably stick with 1.5 for now and will do some performance tests to assess the impact.

ruven commented:
Yes, what you are seeing here makes sense. IIPImage always uses the next highest resolution, but sometimes, this can be only slightly larger than the requested size, so resampling down won't be particularly good. In cases when the next highest resolution is just less than 2x larger than the requested size, then using an even higher source resolution than this from which to resample probably won't make any noticeable difference.

It would be good to find a sweet-spot for this criteria, which could be set as a default in iipsrv for users wanting high quality resampling. 1.5 is already quite reasonable. Have you tried around 1.33? In this case only a third of requests on average would require resampling from a higher resolution.

beaudet commented:
I think 1.5 is a good compromise between quality and performance.

Blowing up the image, I'm able to distinguish between 1.4 and 1.5; it's not plainly obvious, but there are definitely differences.  Quality is much closer between 1.5 and 1.6 although blowing up the pixels still shows some minor differences, although they border on almost imperceptible, particularly when the images are not artificially zoomed.  Below are some examples (the output dimensions had to be adjusted between the cases to ensure different resolutions for the comparison).  You also have to disregard the one pixel stretching that occurs between the images in each case - that stretching is probably due to a vips bug in the PTIF generation, but I still need to look into that more deeply.

case 1: 1.4
![image](https://cloud.githubusercontent.com/assets/77699/26002814/cff6b740-36fe-11e7-9413-990c163b0dd6.png)

case 1: vs 1.5
![image](https://cloud.githubusercontent.com/assets/77699/26002824/d523dc98-36fe-11e7-9970-d402a1252d3a.png)


and 


case 2: 1.5
![image](https://cloud.githubusercontent.com/assets/77699/26002585/2069ee28-36fe-11e7-8e94-8f6aa4488831.png)

case 2: vs. 1.6
![image](https://cloud.githubusercontent.com/assets/77699/26002593/2672d7bc-36fe-11e7-81cb-2cf9fd13b3d2.png)

beaudet commented:
and between 1.5 and 1.6 at lower resolutions where the color intensity of the buttons on the gown are noticeably affected when blowing up the image, but difficult to perceive when not zoomed.  There is also a slight increase in sharpness in the sleeve detail on her left arm in the 1.6 version that is just barely perceptible when not zoomed in, but I really think these differences are really nitpicky and probably not worth upping the ante to 1.6.  1.6 vs. 1.7 has some of the same differences but less perceptible and they can only be seen when the pixels are highly zoomed.

case 3: 1.5
![image](https://cloud.githubusercontent.com/assets/77699/26003151/e5ecf2ac-36ff-11e7-82a6-ad972dc9b877.png)

case 3: vs. 1.6
![image](https://cloud.githubusercontent.com/assets/77699/26003163/ee31fc14-36ff-11e7-8fe4-2df5927b4de5.png)

and 

case 4: 1.6
![image](https://cloud.githubusercontent.com/assets/77699/26003617/38432002-3701-11e7-8d94-6773f458ce71.png)

case 4: vs. 1.7
![image](https://cloud.githubusercontent.com/assets/77699/26003629/3df10172-3701-11e7-9b42-014b8e658bbd.png)

beaudet commented:
@ruven, assuming you're ok with 1.5 when a user wants higher quality resizing, how do you want to handle this in the configuration?  I would think you'd want the default to be 1.0 for backwards compatibility just as bilinear is the default algorithm. I also haven't tested the impact on bilinear as closely as with Lanczos since presumably someone wanting higher quality would choose Lanczos prior to upping the oversampling factor.   I suppose we could bake in number of pre-defined sets of resizing algorithm and resampling level, e.g.

RESIZE_QUALITY: BEST, BETTER, GOOD, AVERAGE, LOW, LOWER, LOWEST 

Which, if present, ignores the algorithm and resampling factor and causes something like the following to be applied:

BEST: Lanczos, resampling 2.0
GREAT: Lanczos, resampling 1.5
ENHANCED: Lanczos: resampling 1.33
BETTER: Lanczos, resampling 1.0
GOOD: bilinear, resampling 1.5
NORMAL: bilinear, resampling 1.33
REDUCED: bilinear, resampling 1.0
LOW: nearest neighbor, resampling 1.5
LOWER: nearest neighbor: resampling 1.33
WORST: nearest neighbor, resampling 1.0

ruven commented:
Yes, having presets as well as precise control would definitely be a good idea. We can just call this "resampling quality" or something. I guess we can just have these as just numbers from 0-9 rather than words.  For the factors themselves, maybe calling them something more like thresholds would be better? Because in reality you wouldn't be resampling from an image 1.5 x larger, but from the next highest existing resolution that is at least 1.5 x larger. So, it's acting more as a threshold. Any thoughts?

Having 1.33 bilinear for the default sounds like good compromise.

For the lower qualities, does nearest neighbor at 1.5 or 1.33 make any visible difference?

beaudet commented:
Sounds good.  I'm not sure about the nearest neighbor differences.  I'll have to work that out when I add it as an option.  Once you've had a chance to accumulated the PRs submitted so far into the develop branch, I'll clone it and submit some new PRs against develop for this work.

ruven commented:
I'm going to integrate your oversampling code into the master branch, but I think it should be implemented differently. Basically this should all happen in the View class and be completely transparent to the CVT output code. The function _calculateResolution()_ in View.cc is what determines the appropriate source resolution to use. The code is simply this:

```
// Calculate the resolution number for this request                                                                                       
  while( d >= rs ){
    d = d/2;
    j++;
  }
```

So, instead of d >= rs, we just need d >= rs*oversampling.

And the term oversampling is perhaps misleading. It's really a minimum downsampling factor. If you request an image of size 501x501 pixels and have existing resolutions within the TIFF of 500x500 and 1000x1000, the current implementation will use the 1000x1000 resolution, which is already an oversampling of 1.9999. If you request 499x499, it will currently use 500x500. With an "oversampling" of, say 1.5, 501x501 will still be downsampled from 1000x1000, but 499x499 will be downsampled from 1000x1000 rather than 500x500.

For your above examples, what was the requested size relative to the existing resolution size? To find the sweet spot properly, you should get an original high resolution image and use Gimp, Vips or Photoshop etc to resize using Laczos3 down to different sizes, say: 1100, 1200, 1400, 1500, 1600, 1700, 1800 pixels, then for each of those do a second resize using lanczos3 from 1200x1200 etc down to exactly 1000x1000 and compare those series of 1000x1000 resized images. It's very subjective of course, but you should try yourself and we can compare the numbers we come up with.

ruven commented:
I've just done some experiments with different "oversampling" ratios. Rather than use a subjective comparison, I've used SSIM (structural similarity) to compare a little more objectively the quality of 2 images (the higher the SSIM the better). I've created a reference downsampled using Lanczos3 directly from a high resolution image, then done the 2 step resizing I mentioned above. I used imagemagick for all the resizing. The result is quite interesting and shows that the image quality plateau's off after an oversampling of around 1.4 and that going beyond 2 makes a negligible difference:

![resizing_graph](https://user-images.githubusercontent.com/776628/27822092-bbdc2144-60a4-11e7-8814-856b0bc814ce.png)

So, perhaps a default minimum oversampling of around 1.4 could be used by iipsrv, which isn't such an onerous overhead.

beaudet commented:
Yes, I realized that I had put the resizing in the wrong place and moved it there as well but I didn't have a chance to commit that change I guess.  Anyway, sounds great.  Regarding 1.4 as a minimum, that will most definitely force IIP to resample with CVT each time rather than serving up tiles.  I think that's desirable actually because it will provide more of a continuum in terms of quality, although the differences should now be negligible with the recent quality increases and updates to VIPS.   So, yeah, I agree with using 1.4 as a minimum.  Interesting analysis with SSIM.  I'll read up on that.

ruven commented:
> Regarding 1.4 as a minimum, that will most definitely force IIP to resample with CVT each time rather than serving up tiles

Not really. It means that if the next highest resolution is less than 1.4 times larger (for example you ask for a 360px pixel image and there exists a resolution at 500px - ratio 1.39), it will ask for a higher resolution. So, on average only 40% of resize requests will need to go through this.

Looking at other resizing algorithms, however, there is a difference in where this sweet spot is. Lanczos2 and Catmull-Rom need something nearer 1.5 whereas both Mitchell and Bilinear need around 2 (or more).

![resizing_graph](https://user-images.githubusercontent.com/776628/27953922-f16ee07e-630d-11e7-8592-622aa0386c83.png)

And combining all this with the recent rescaling updates to Vips, we'll have very high quality output indeed. I'll do an end-to-end quality comparison to imagemagick or vips to quantify this. It won't be quite as good, of course, but the resizing in iipsrv is done in milliseconds rather than 10's of seconds :wink:

beaudet commented:
Yeah, I meant to suggest that CVT will / should?
 ignore the optimization that returns the native tile from the PTIF if the request happens to align with the native tile size.

ruven commented:
If the request aligns with the native file resolution, then using this directly will always provide the best quality output. Especially now that with Vips each resolution level has already been resized with Lanczos3.

beaudet commented:
hopefully the differences between the native PTIF and the resampled cases will be undetectable.  My only concern is that if a client is generating a spectrum of similar sizes, that there could be a detectable quality difference in the middle of that spectrum - that was pretty obvious with the old methods of resampling and I suppose that will happen anyway near a 1.4 threshold - I just hope the differences aren't as noticeable.  It makes total sense to return those in a zoom viewer context of course (for performance and quality reasons) - but I can imagine cases where someone would want to force resampling all the time rather than returning tiles.   For example, if I have a page of thumbnails constrained to a bounding box I would want all of the thumbnails to be similar in quality - ideally as high quality as possible - but for there not to be noticeable differences in quality between them.  Since the PTIF page resolutions are a function of the tile size and the original image dimensions, we don't have much control over the frequency of these JTL hits.  Does that make sense or do you think I'm splitting hairs now?

ruven commented:
> but I can imagine cases where someone would want to force resampling all the time rather than returning tiles

For the exact alignment case, resampling something that has already been properly resampled won't improve the quality. In fact it will even reduce the quality. In the graphs, the Lanczos3 resampling never goes to 1 (perfect reproduction) even at very high oversampling. Resampling something twice will result in  worse quality than doing the same thing in a single sweep.

For transitions, this will depend where the user sets the threshold - I'll make this configurable. But for Lanczos3, for example, there's negligible difference in SSIM quality beyond 1.4, so in theory this shouldn't be visible. You could do more subjective visual tests to validate this. I'll do myself more SSIM tests to compare the final iipsrv output to imagemagick output to see if there's a visible transition for a fixed oversampling ratio.

beaudet commented:
What tool are you using for calculating SSIM by the way?

ruven commented:
There's a python implementation called [pyssim](https://github.com/jterrace/pyssim)
